### DISCUSSION TITLE
A PKM System with Quarto

### DISCUSSION THREAD

**Discussion by neeldhara** [quarto-dev/quarto-cli#6789] [Show and Tell]
### Description

For those who are not in PKM-land, PKM = Personal Knowledge Management, and of late there's been a lot of interest in tools designed to manage personal knowledge - these tend to be either super-visual/WYSIWYG (e.g, Notion, Craft, etc.) or markdown-based (e.g, Obsidian, Logseq, etc). 

I've been going back and forth between a few tools, with my personal priorities being:

- local-first
- speedy capture and render
- support for inline LaTeX
- support for inline hand-drawn images (esp. if there's an iPad app)
- support for infinite canvas whiteboarding
- support for backlinks, databases, tags
- support for publishing a part of the KB to a blog, or to a PDF

For example, as someone who teaches undergraduate courses frequently, a common use-case for me is the preparation of assessments (quizzes, homework, exams, etc). I mostly use the exam class in LaTeX to generate these. I am aware that LaTeX has some packages that let you build a question bank, but this does not come with a nice UI to filter and sort though problems based on either difficulty or theme.

A lot of the popular PKM tools will allow me to do this with a nice UI, as will several LMS-es, but then I have a lot of duplicate work as I port them over into my LaTeX files.

Meanwhile, I have been using Quarto for my website for a while now, and I run a puzzles group, and it's been pretty seamless for me to maintain my collection of puzzles and have them filtered by topic: and it looks great!

So I setup a local repo for tracking exercises, and so far it's working out quite nicely:

![CleanShot 2023-09-10 at 15 45 06](https://github.com/quarto-dev/quarto-cli/assets/15953732/630c3361-e107-4a96-9240-546f34b2c017)

By using an appropriate filter, I am able to generate a PDF of an exam by including the relevant questions on a separate quiz page via Quarto's support for including files. I do have some details that I am still figuring out, since I am not really familiar with how pandoc/lua filters work, but I hope to sort them out in due course, likely with help from here :)

[This is the PDF file](https://www.neeldhara.com/courses/2023/03-es214/exams/qp1-solutions.pdf) that was produced from this setup, and I am quite happy with it :)

I am slowly transitioning my projects into this system as well:

![CleanShot 2023-09-10 at 15 46 03](https://github.com/quarto-dev/quarto-cli/assets/15953732/700758c8-df84-48aa-811f-193c561cbcd5)

I maintain a separate list of papers (I'm still working on automating some parts - I think a script can transform BIB to YAML, and that would be nice), and I tag each paper with the projects that they are relevant to, and they get pulled up in the project page with a custom listings.ejs setup like so:

```{=html}

<table class="quarto-listing-table table">
    <thead>
        <tr>
            <td>Title</td>
            <td>Hook</td>
            <td>PDF</td>
        </tr>
    </thead>
    <tbody class="list">
    <% 
    for (const item of items) {
        if(item.projects){
        for (const p of item.projects) {  
            if(p.project == templateParams.pid) { %>
                <tr data-index="0" data-categories="null" <%= metadataAttrs(item) %>>
                    <td class="listing-title">
                    <a href="<%= item.path %>"><%= item.title %></a>
                    </td>
                    <td class="listing-hook">
                    <a href="<%= item.hook %>">&nbsp;&nbsp;🔗</a>
                    </td>
                    <td class="listing-asset">
                    <a href="<%= item.asset %>">&nbsp;&nbsp;📝</a>
                    </td>
                </tr>
            <% } 
        }
        }
    } 
    %>
</tbody>
</table>
```

I extensively use Hookmark URLs in the metadata, so I can open relevant files immediately - whether it's PDFs, videos, or the qmd files themselves - if I am looking at a note that I want to edit, it's super convenient to do it by clicking on the hookmark URL that's a part of all the templates.

Since quarto render on individual files works incrementally, I hope this will scale up well - I anticipate adding hundreds, and eventually perhaps thousands of files (I am throwing in time logs, journal entries, and the rest of my life into this gradually). 

The fact that I can customize exactly how it looks is a big win for me, as is the fact that I can do reliable PDF publishing without having to duplicate my work. Since my website is also on Quarto with the same theme, I anticipate doing rough work on my blogs and lecture notes within the personal local repo, and transitioning without too much difficulty once things look good.

As of now, I am not sure how to work out access on mobile devices, infinite canvas, and handwritten notes. However, the fact that I output HTML implies that there's promise in terms of being able to perhaps embed things I need. As for mobile, I plan to use Drafts for quick capture... also, if I just want to see my notes, I think I can do a protected server deploy on Netlify eventually, but I am yet to work this out.

I just wanted to give a shout out to the Quarto team - it's a wonderful tool that is superbly documented! ❤️

**Comment by the-solipsist** [comment]
> I think a script can transform BIB to YAML, and that would be nice

One way to transform a .bib file into a .yaml file is to run it through pandoc:
`pandoc -f biblatex <filename.bib> -t markdown -o <output-filename.yaml>`

-----
A question: Am I right in assuming that you modified an existing ejs listing template?  If so, which one?  I could only find https://github.com/quarto-dev/quarto-cli/blob/53bf5f4b8b50b6c8d28b80a01cee012626c180a4/src/resources/projects/website/listing/listing-table.ejs.md?plain=1 , which looks very different.  I can sort of understand your template's code, but can't quite understand how `listing-table.ejs.md` functions.

**Reply by mcanouil** [reply]
@the-solipsist Listing templating is documented at https://quarto.org/docs/websites/website-listings-custom.html

**Reply by neeldhara** [reply]
> > I think a script can transform BIB to YAML, and that would be nice
> 
> One way to transform a .bib file into a .yaml file is to run it through pandoc: `pandoc -f biblatex <filename.bib> -t markdown -o <output-filename.yaml>`

Ah, great idea - I could not resist following up on this :D So I now have the following workflow:

- I download a .bib file anywhere on the internets
- I wait for ~5 seconds
- A new browser tab opens up corresponding to the .qmd file that contains the details of the paper in the meta data and a basic empty template.

This is what it looks like (I only click twice in this entire recording - once to download the .bib file, and the other time to refresh the page that came up):

https://github.com/quarto-dev/quarto-cli/assets/15953732/bbec918b-6317-4899-ac4b-e35d6357697a

Behind the scenes:

- The .bib file is downloaded to ~/Downloads
- Hazel, which is monitoring my Downloads folder, sees a new file with a ".bib" extension that was added recently and whose size is < 1000 bytes
- Hazel moves this file to ~/Things/Inbox/paper
- Keyboard Maestro is watching the ~/Things/Inbox/paper folder - it sees a new item, gets triggered to rename it to `newpaper.bib` and moves this to the correct folder in my Quarto repo, where it runs the pandoc script and a little bit of other housekeeping (to give the file an appropriate ID; append the contents of the bib file to an `all.bib` file, delete the `newpaper.bib` file, etc etc), and runs the quarto preview command and opens up the URL in a new browser tab.

> A question: Am I right in assuming that you modified an existing ejs listing template? If so, which one? I could only find https://github.com/quarto-dev/quarto-cli/blob/53bf5f4b8b50b6c8d28b80a01cee012626c180a4/src/resources/projects/website/listing/listing-table.ejs.md?plain=1 , which looks very different. I can sort of understand your template's code, but can't quite understand how `listing-table.ejs.md` functions.

As @mcanouil pointed out, I am using a custom .ejs setup. I think you can take the data in listings and combine and showcase them as you like: anywhere from a plain listing to a gallery, and so on. Also since .ejs (apparently) supports scripting, you can, for example, apply different styling to different kinds of lists (I do this for the icon things on my [puzzles page](https://www.neeldhara.com/tgif/) - they are different based on category, and you can find the source ejs file for that [here](https://github.com/neeldhara/quartosite/blob/7441300513ca12eb732b62282a4baf82412bff6b/tgif/tgif.ejs#L4)).

Thanks for the tip on the script! 🙌

**Comment by MichelNivard** [comment]
Hi @neeldhara I am trying to set up something similar, I have a fairly basic question. Is your PKM basically a massive quarto website or blog project?  Or do you use a default project or a custom project template you developed yourself?

**Reply by fredguth** [reply]
@neeldhara, Can you share the repo?

**Comment by fredguth** [comment]
Have you noticed in Quarto homepage the mention to "Knowledge Repos"? 

I wonder if there is a repo template for a Knowledge Repo using Quarto.  I am in the process of creating a Knowledge Base that I see the need to integrate with Quarto.
